## Interpreting supervised machine learning inferences in population genomics using haplotype matrix permutations
In this work, we introduce a systematic permutation approach that progressively disrupts population genetics features within input test haplotype matrices, including linkage disequilibrium, haplotype structure, and allele frequencies.
By measuring performance degradation after each permutation, the importance of each feature can be assessed.

### Preprint
[bioRxiv](https://www.biorxiv.org/content/10.1101/2025.03.24.644668v1)

### Reproduce

To reproduce the result for each of the three CNNs evaluated in this work, use scripts provided in the 3 respective subdirs. Each subdir includes further instruction for reproducing each case.

### Extra: Conference presentations of this work
* [TAGC24 Poster](https://github.com/lntran26/lntran26.github.io/blob/4e461eaf627614b75ec47d9a8f72fd5491880fb9/files/TAGC_24_Tran_final.pdf)
* [ISMB24 MLCSB COSI Talk](https://github.com/lntran26/lntran26.github.io/blob/1a5497584df962c9d5643b013523ecaf0fa8c2ef/files/ISMB24.pdf)
* [EVO-WIBO Poster](https://github.com/lntran26/lntran26.github.io/blob/master/files/Tran_ConfuseNN_final.pdf)
